metrics/metrics.go

```diff
@@ -0,0 +1,192 @@
+	driverIssueCallCount                 *prometheus.CounterVec
+	driverIssueErrorCount                *prometheus.CounterVec
+	managedVolumeCount                   *prometheus.CounterVec
+	managedCertificateCount              *prometheus.CounterVec
+}
+
+// New creates a Metrics struct and populates it with prometheus metric types.
+func New(logger *logr.Logger) *Metrics {
+	var (
+		certificateRequestExpiryTimeSeconds = prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "certificate_request_expiration_timestamp_seconds",
+				Help:      "The date after which the certificate request expires. Expressed as a Unix Epoch Time.",
+			},
+			[]string{"name", "namespace", "issuer_name", "issuer_kind", "issuer_group"},
+		)
+
+		certificateRequestRenewalTimeSeconds = prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "certificate_request_renewal_timestamp_seconds",
+				Help:      "The number of seconds before expiration time the certificate request should renew.",
+			},
+			[]string{"name", "namespace", "issuer_name", "issuer_kind", "issuer_group"},
+		)
+
+		certificateRequestReadyStatus = prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "certificate_request_ready_status",
+				Help:      "The ready status of the certificate request.",
+			},
+			[]string{"name", "namespace", "condition", "issuer_name", "issuer_kind", "issuer_group"},
+		)
+
+		driverIssueCallCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "driver_issue_call_count",
+				Help:      "The number of issue() calls made by the driver.",
+			},
+			[]string{"node", "volume"},
+		)
+
+		driverIssueErrorCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "driver_issue_error_count",
+				Help:      "The number of errors encountered during the driver issue() calls.",
+			},
+			[]string{"node", "volume"},
+		)
+
+		managedVolumeCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "managed_volume_count",
+				Help:      "The number of volume managed by the csi driver.",
+			},
+			[]string{"node"},
+		)
+
+		managedCertificateCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "managed_certificate_count",
+				Help:      "The number of certificates managed by the csi driver.",
+			},
+			[]string{"node"},
+		)
+	)
+
+	// Create Registry and register the recommended collectors
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+		collectors.NewGoCollector(),
```

Comment:
I feel like these *should* be auto-registered, but I think the `Metrics` struct itself should be able to be setup so that it *doesn't* embed these Go metrics into them (i.e. so it only adds in the csi-lib specific metrics to the registry).

Perhaps we need some 'basic' constructor function for the csi-lib metrics, with some form of `NewRecommendedMetrics` function that calls the basic one, but also adds in standard Go-esque metrics? Not sure :)

Reply:
Agree, I will remove those default registry and let the caller to pass their Registry instead.